community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/planner/PlannerQuery.scala

```diff
@@ -101,6 +101,11 @@ sealed trait PlannerQuery {
     case Some(tailQuery) => this.updateTail(_.updateTailOrSelf(f))
   }
 
+  def tailOrSelf: PlannerQuery = tail match {
+    case None => this
+    case Some(t) => t.tailOrSelf
```

Comment:
So this is 'recursively' checking for the _final_ tail, right? I think the name is a bit misleading; how about `lastTail`?